### PR TITLE
perpetual: new module for mapping value types into memory

### PR DESCRIPTION
**Pull request by sdegtiarev**
I rewrote std.perpetual module to be based on top of std.mmfile.
It becomes clear that high-level data handling should be separate from low-level shared memory allocation. In addition to architectural advantages, this also clearly answers the question: what new the perpetual module added to existing functionality.

### CODE REVIEW

**Comment by quickfur** [comment]
Why `std.perpetual` instead of `std.persistence`? Persistence is well-established terminology for what this module does, "perpetual" sounds strange.

File: std/experimental/perpetual/perpetual.d [hunk 1]
```diff
@@ -0,0 +1,192 @@
+module std.perpetual;
```
Comment (quickfur) [inline std/experimental/perpetual/perpetual.d]: Modules need to have proper ddoc header.

File: std/experimental/perpetual/perpetual.d [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+		private Element[] _value;
+		enum bool dynamic=true;
+		static assert(!hasIndirections!Element, Element.stringof~" is reference type");
+		@property Element[] Ref() { return _value; }
+		string toString() { return to!string(_value); }
+
+	}
+	else
+	{
+	// value type
```
Comment (quickfur) [inline std/experimental/perpetual/perpetual.d]: What about class types?

**Comment by quickfur** [comment]
Also, please conform to the Phobos style guide: http://dlang.org/dstyle.html (note especially the last section on Additional Requirements for Phobos).

File: std/experimental/perpetual/perpetual.d (hunk 2 above)
Comment (sdegtiarev) [inline std/experimental/perpetual/perpetual.d]: On Thu, 8/13/15, Dmitry Olshansky <notifications@github.com> wrote:
> the name most likely gonig to change as persistence is more general then mmap-ing objects 
>  to files.
Just accepted advise, I'm for persistent myself

Comment (sdegtiarev) [inline std/experimental/perpetual/perpetual.d]: @quickfur 
> What about class types?
Class types are references, unfortunately you can't store them neither pointers.
That is what serialization does, but it is very different use case.

Comment (quickfur) [inline std/experimental/perpetual/perpetual.d]: OK, fair enough.

**Comment by quickfur** [comment]
Since this PR is proposing to add a new Phobos module, I think it should at least be discussed on the forum and go through the Phobos review process.

**Comment by sdegtiarev** [comment]
@quickfur  Certainly, I still need to fix some urgent issues, plus, a little more detailed documentation would be useful

File: std/experimental/perpetual/perpetual.d (hunk 1 above)
Comment (DmitryOlshansky) [inline std/experimental/perpetual/perpetual.d]: module std.experimental.perpetual;

**Comment by DmitryOlshansky** [comment]
Another thing to note - get rid of merge commits and use git rebase to squash minor edits and things like moving files around.

**Comment by JackStouffer** [comment]
Why are there two PRs for the same thing? https://github.com/D-Programming-Language/phobos/pull/3562

**Comment by sdegtiarev** [comment]
This is final version, the code is ready for review and discussion

**Comment by JackStouffer** [comment]
Please close the other PR then.

**Comment by JackStouffer** [comment]
Now, to start the review process, you need to do two things:

1. Please create a form thread in general asking for comments from the community: http://forum.dlang.org/group/general

2. Please add your code to the bottom of the review queue: http://wiki.dlang.org/Review_Queue

Also, in your code, there is a lot of extra whitespace. There should never be more than one line of whitespace between lines of documentation and code.

File: std/experimental/perpetual.d [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+		static assert(!hasIndirections!Element
+		    , Element.stringof~" is reference type");
+
+		/** Proxy to underlying dynamic array type.
+		 *  Forwards calls to opSlice(), length() etc
+		 */
+		@property Element[] Ref() { return _value; }
+
+		/// Convert to string the value, not the class itself
+		@property toString() { return to!string(_value); }
```
Comment (JackStouffer) [inline std/experimental/perpetual.d]: This code unnecessarily allocates. Please fix by using the overload shown in http://dlang.org/phobos/std_format.html#.formatValue.10

This applies to all `toString()` functions in this code.

File: std/experimental/perpetual.d [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+		auto p5=Perpetual!(char[32])(file[5]);
+		p5="hello world";
+
+		// second order static array
+		auto p8=Perpetual!(char[3][5])(file[6]);
+		p8[]="..."; p8[1]="one"; p8[2]="two";
+
+
+		/// Compile time errors
+		// Perpetual!(char*)("?");        //ERROR: "char* is reference type"
```
Comment (JackStouffer) [inline std/experimental/perpetual.d]: I believe you can test this via `static assert(!is(typeof(Perpetual!(char*)("?"))));` This applies the others as well.

File: std/experimental/perpetual.d [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+ 	
+
+ // get reference to wrapped object.
+	alias Ref this;
+
+}
+
+
+///
+unittest {
```
Comment (JackStouffer) [inline std/experimental/perpetual.d]: This is a really long ddoc. Please consider breaking this into multiple unit tests. 

Also, please put the `{` on its own line.

File: std/experimental/perpetual.d [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+
+}
+
+
+
+
+
+
+
+private import std.exception;
```
Comment (JackStouffer) [inline std/experimental/perpetual.d]: Please put all imports at the top of the file.

File: std/experimental/perpetual.d [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+				{
+					_master=false;
+					_writeable=true;
+					return _fd;
+				}
+
+				errnoEnforce(false, path);
+			}
+		}
+		else version(Windows)
```
Comment (JackStouffer) [inline std/experimental/perpetual.d]: What's with all the empty `version(Windows)` blocks?

**Comment by JackStouffer** [comment]
Also, please change all of your tab indentation to four spaces.

**Comment by sdegtiarev** [comment]
Will do, thanks.
I love this, will use: static assert(!is(typeof(..expr..)));
--------------------------------------------
On Sat, 9/19/15, Jack Stouffer <notifications@github.com> wrote:

 Subject: Re: [phobos] perpetual: new module for mapping value types into memory (#3625)
 To: "D-Programming-Language/phobos" <phobos@noreply.github.com>
 Cc: "Sergei Degtiarev" <sdegtiarev@yahoo.com>
 Date: Saturday, September 19, 2015, 1:04 PM
 
 Now, to start the review process,
 you need to do two things:
 
 
 Please create a form thread in general asking for
 comments from the community: http://forum.dlang.org/group/general
 Please add your code to the bottom of the review
 queue: http://wiki.dlang.org/Review_Queue
 
 
 Also, in your code, there is a lot of extra whitespace.
 There should never be more than one line of whitespace
 between lines of documentation and code.
 
 —
 Reply to this email directly or view
 it on GitHub.

File: std/experimental/perpetual.d (hunk 7 above)
Comment (sdegtiarev) [inline std/experimental/perpetual.d]: Still need to filled. I'm not very good with Windows system calls, neither have a machine to compile and test, so I left it to the last minute and forgot.
--------------------------------------------
On Sat, 9/19/15, Jack Stouffer <notifications@github.com> wrote:

 Subject: Re: [phobos] perpetual: new module for mapping value types into memory (#3625)
 To: "D-Programming-Language/phobos" <phobos@noreply.github.com>
 Cc: "Sergei Degtiarev" <sdegtiarev@yahoo.com>
 Date: Saturday, September 19, 2015, 3:13 PM
 
 In std/experimental/perpetual.d:
 > +					return _fd;
 > +				}
 > +
 > +				/// otherwise, try to open read/write existing
 file
 > +				_fd=.open(toStringz(path), O_RDWR,
 S_IRUSR|S_IWUSR);
 > +				if(_fd >= 0)
 > +				{
 > +					_master=false;
 > +					_writeable=true;
 > +					return _fd;
 > +				}
 > +
 > +				errnoEnforce(false, path);
 > +			}
 > +		}
 > +		else version(Windows)
 
 What's with all the empty
 version(Windows) blocks?
 
 —
 Reply to this email directly or view
 it on GitHub.

File: std/experimental/.gitignore [hunk 8]
```diff
@@ -0,0 +1,7 @@
+
+*.o
+perpetual
+core
+Makefile
+perpetual.html
+
```
Comment (greenify) [inline std/experimental/.gitignore]: Is this file needed once it would be in Phobos? All the other modules are fine without this.

Comment (sdegtiarev) [inline std/experimental/.gitignore]: No, of course not.

**Comment by JackStouffer** [comment]
I would suggest this be closed until this module is ready for prime time. Right now all it is doing is using autotester time needlessly.

**Comment by sdegtiarev** [comment]
It is actually ready, I don't see what else I can do with it.
Changed status to "voting"

File: std/experimental/perpetual.d [hunk 9]
```diff
@@ -0,0 +1,725 @@
...
+	        sink.formatValue(_value, fmt);
+        }
+
+        /// Index of uninitialized part of the array
+        @property initIndex() const { return _heap.unplowed/Element.sizeof; }
+    }
+    // scalar value or static array
+    else
+    {
+        alias Element=T;
```
Comment (greenify) [inline std/experimental/perpetual.d]: nitpick: whitespace

File: std/experimental/perpetual.d [hunk 10]
```diff
@@ -0,0 +1,725 @@
...
+        private Element* _value;
+        private enum bool dynamic=false;
```
Comment (greenify) [inline std/experimental/perpetual.d]: nitpick: whitespace

File: std/experimental/perpetual.d [hunk 11]
```diff
@@ -0,0 +1,725 @@
...
+        static assert(!hasIndirections!T
+            , T.stringof~" is reference type");
+
+        /** Reference to underlying scalar value
+         */
+        @property ref Element Ref() {    return *_value; }
```
Comment (greenify) [inline std/experimental/perpetual.d]: too much whitespace

File: std/experimental/perpetual.d [hunk 12]
```diff
@@ -0,0 +1,725 @@
...
+{
+    // underlying shared memory allocator
+    private ShMem _heap;
+    private enum _tag="Perpetual!("~T.stringof~")";
+
+    //dynamic array
+    static if(is(T == Element[],Element))
+    {
+        private Element[] _value;
+        private enum bool dynamic=true;
```
Comment (greenify) [inline std/experimental/perpetual.d]: nitpick: whitespace

File: std/experimental/perpetual.d [hunk 13]
```diff
@@ -0,0 +1,725 @@
...
+
+    @property bool writeable() const { return _writeable; }
+
+    @property bool master() const { return _master; }
+
+    @property auto unplowed() const { return _index; }
+
+    
+
+
```
Comment (greenify) [inline std/experimental/perpetual.d]: nitpick: only one line between methods

**Comment by greenify** [comment]
> It is actually ready, I don't see what else I can do with it.
> Changed status to "voting"

@sdegtiarev can you please make the NG annoucement then?
@JackStouffer do you volunteer as a review manager for the voting?
(don't forget to update [the review queue](https://wiki.dlang.org/Review_Queue) based on the voting)

**Comment by JackStouffer** [comment]
Sorry, I don't have time right now.

**Comment by wilzbach** [comment]
>> Changed status to "voting"
> @sdegtiarev can you please make the NG annoucement then?

I wouldn't mind handling the review - so move this forward and start the discussion ;-)

**Comment by sdegtiarev** [comment]
I started discussion thread.
https://forum.dlang.org/post/stesmdubyyldrseepqqz@forum.dlang.org
Hopefully, it will push things forward.
--------------------------------------------
On Tue, 5/10/16, Sebastian Wilzbach <notifications@github.com> wrote:
 I wouldn't mind handling the review - so move this
 forward and start the discussion ;-)

**Comment by wilzbach** [comment]
@sdegtiarev 
1) do you mind running "dfmt" on your module? You miss a lot of whitespaces etc.
2) can you squash your commits into one?

@klickverbot i will stay neutral as review manager - my point is that such pending PRs don't help anyone. Either dub or Phobos, but not pending ;-)

**Comment by mleise** [comment]
What happens with pointers in the embedded objects when the file is restored from file or used by multiple processes?

**Comment by sdegtiarev** [comment]
No, pointers are not of course allowed in stored data, this is general limitation for such code. You have to use serialization if you need such functionality.
--------------------------------------------
On Wed, 5/11/16, Marco Leise <notifications@github.com> wrote:
 
 What happens with pointers in the
 embedded objects when the file is restored from file or used
 by multiple processes?

**Comment by sdegtiarev** [comment]
Believe it or not, I can't build dfmt... One of my machines has 2Gb ram (+swap) and dmd reports insufficient memory, another one has 8Gb ram but version dmd-2.67 and reports compilation error: template 'AliasSeq' is not defined. Incredible.
--------------------------------------------
On Wed, 5/11/16, Sebastian Wilzbach <notifications@github.com> wrote:

 Subject: Re: [dlang/phobos] perpetual: new module for mapping value types into memory (#3625)
 To: "dlang/phobos" <phobos@noreply.github.com>
 Cc: "Sergei Degtiarev" <sdegtiarev@yahoo.com>, "Mention" <mention@noreply.github.com>
 Date: Wednesday, May 11, 2016, 6:40 AM
 
 @sdegtiarev 
 
 1) do you mind running "dfmt" on your module? You
 miss a lot of whitespaces etc.
 
 2) can you squash your commits into one?
 
 @klickverbot i will
 stay neutral as review manager - my point is that such
 pending PRs don't help anyone. Either dub or Phobos, but
 not pending ;-)
 
 —
 You are receiving this because you were mentioned.
 Reply to this email directly or view
 it on GitHub

**Comment by wilzbach** [comment]
> Believe it or not, I can't build dfmt... One of my machines has 2Gb ram (+swap) and dmd reports insufficient memory, another one has 8Gb ram but version dmd-2.67 and reports compilation error: template 'AliasSeq' is not defined. Incredible.

Have you tried with a recent dmd version like 2.070 or 2.071?

**Comment by sdegtiarev** [comment]
dmd 2.07 is as memory greedy as as older version. Will try to install and build on more powerful machine tomorrow.
--------------------------------------------
On Sun, 5/15/16, Sebastian Wilzbach <notifications@github.com> wrote:
 
 Have you tried with a recent dmd version like 2.070 or
 2.071?
